our-paxos/paxos.py

```diff
@@ -193,43 +198,110 @@ def proposer(config, id):
-                print("Learned: ", value)
-                learned+=1
-            messages.append(value)
-
-        # If id is < len(msg) we swap -1 (value of not received message) with the received value,
-        # then if id == learned, then we can print the value as learned until we reach first
-        # undefined message
-        elif id < len(messages):
-            messages[id] = value
-            if id == learned:
-                while(messages[learned] != -1 and len(messages) > learned):
-                    print("Learned: ", value)
-                    learned += 1
-
+        inst_id = msg[0]
+        if len(msg) > 1:
+            # If we receive un "update message" from another learner
+            if msg[1] == 1:
+                while inst_id <= len(messages):
+                    messages.append(-1)
+                if messages[inst_id] == -1:
+                    inst_id[inst_id] = msg[3]
+
+            # If we received PAXOS round 2 message
+            if(msg[1] == 2):
+                # print("Instance: {} Learned: {}".format(inst_id, msg[3]))
+                value = msg[3]
+                # If id is > len(msg) we need to extend array of messages up to the needed size,
+                # we don't show that we learned anything, bc we've extended array without assigning
+                # the message with the smaller id
+                if inst_id > len(messages):
+                    print("Init mor: {}".format(inst_id))
+                    while(inst_id - 1 > len(messages)):
+                        messages.append([])
+                        messages_running.append(False)
+                    messages.append([value])
+                    messages_running.append(True)
+                # If id is == len(msg) we apppend a message to the array,
+                # we don't show that we learned anything if learned isn't equal to msg len
+                # because in this case we're still missing message somewhere in the middle
+                elif inst_id == len(messages):
+                    print("Init eq: {}".format(inst_id))
+                    # if learned == len(messages) and len(messages[id]) == QUORUM_AMOUNT - 1:
+                    #     if value == messages[id][0]:
+                    #         print("Instance: {} Learned: {}".format(id, value))
+                    #     learned += 1
+                    messages.append([value])
+                    messages_running.append(True)
+
+                # If id is < len(msg) we swap -1 (value of not received message) with the received value,
+                # then if id == learned, then we can print the value as learned until we reach first
+                # undefined message
+                elif inst_id < len(messages):
+                    # print("Init les: {}, msg: {}, learned: {}, quorum: {}".format(inst_id, messages[inst_id], learned, QUORUM_AMOUNT))
+                    if len(messages[inst_id]) >= QUORUM_AMOUNT - 1 and inst_id == learned:
```

Comment:
I am not sure about this condition. 

Let's say that the learner has received a majority of messages for instance 0, 2, 3 in order so it has never met this condition before. The state should be: 
```python
messages = [ [val, val, val], [ ], [val, val, val], [val, val, val] ]
learned = 0
```
Then it receives 2 values for instance 1 and enters this condition:
```python
2 >= 2 #true - quorum reached
1 == 0 #false
```
So it never enters here and never will because `learned` is updated only inside of this condition, as far as I can see?

Reply:
Yes, but learned would be 1 at this point, bc we already learned `messages[0]` and made `learned += 1`

Reply:
So the thing in here is that `learned won't be 0 at this point`